src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/create/actions/RevertFlowMirrorPathAllocationAction.java

```diff
@@ -66,21 +67,24 @@ protected void perform(State from, State to, Event event, FlowMirrorPointCreateC
             Flow flow = getFlow(stateMachine.getFlowId());
             PathId oppositePathId = flow.getOppositePathId(stateMachine.getFlowPathId()).orElse(null);
             Set<PathId> involvedPaths = newHashSet(stateMachine.getFlowPathId(), oppositePathId);
-            DataAdapter dataAdapter = new PersistenceDataAdapter(persistenceManager, involvedPaths, newHashSet(),
+            Set<PathId> involvedMirrorPaths = newHashSet(stateMachine.getForwardMirrorPathId());
+            DataAdapter dataAdapter = new PersistenceDataAdapter(persistenceManager, involvedPaths, involvedMirrorPaths,
                     newHashSet(stateMachine.getMirrorSwitchId()), false);
-            stateMachine.getRevertCommands().addAll(ruleManager.buildMirrorPointRules(mirrorPoints.get(), dataAdapter));
+            FlowMirrorPath forwardMirrorPath = getFlowMirrorPath(stateMachine.getForwardMirrorPathId());
+            stateMachine.getRevertCommands().addAll(ruleManager.buildMirrorPointRules(forwardMirrorPath, dataAdapter));
         } else {
             log.warn("Can't find mirror points for flow path {} and mirror switch {}. May cause excess rules.",
                     stateMachine.getFlowPathId(), stateMachine.getMirrorSwitchId());
         }
 
-        resourcesManager.deallocateCookie(stateMachine.getUnmaskedCookie());
+        resourcesManager.deallocatePathResources(stateMachine.getFlowResources());
         flowMirrorPathRepository.remove(stateMachine.getForwardMirrorPathId());
```

Comment:
All these removals must be a single atomic operation? What would happen if it fails on the repository removal? It is already de-allocated, but the state machine still holds the reference?